src/EFCore.PG/Query/Internal/NpgsqlSqlTranslatingExpressionVisitor.cs

```diff
@@ -56,17 +57,38 @@ public override SqlExpression TranslateCount(Expression expression = null)
                     _sqlExpressionFactory.Function("COUNT", new[] { _sqlExpressionFactory.Fragment("*") }, typeof(long))),
                 typeof(int), _sqlExpressionFactory.FindMapping(typeof(int)));
 
-//            // In PostgreSQL SUM() doesn't return the same type as its argument for smallint, int and bigint.
-//            // Cast to get the same type.
-//            // http://www.postgresql.org/docs/current/static/functions-aggregate.html
-//            if (sqlFunctionExpression.FunctionName == "SUM")
-//        {
-//            if (sqlFunctionExpression.Type == typeof(int))
-//                Sql.Append("::INT");
-//            else if (sqlFunctionExpression.Type == typeof(short))
-//                Sql.Append("::SMALLINT");
-//            return expr;
-//        }
+        // In PostgreSQL SUM() doesn't return the same type as its argument for smallint, int and bigint.
+        // Cast to get the same type.
+        // http://www.postgresql.org/docs/current/static/functions-aggregate.html
+        public override SqlExpression TranslateSum(Expression expression)
+        {
+            if (!(expression is SqlExpression sqlExpression))
+            {
+                sqlExpression = Translate(expression);
+            }
+
+            if (sqlExpression == null)
+            {
+                throw new InvalidOperationException(CoreStrings.TranslationFailed(expression.Print()));
+            }
+
+            var inputType = sqlExpression.Type.UnwrapNullableType();
+
+            // Note that there is no Sum over short in LINQ
+            if (inputType == typeof(int))
```

Comment:
`SUM` has `bigint` as the return type when the argument of type `smallint` or `int`.

Reply:
Right - this is why this exists for int. For smallint, there is actually no .NET overload of Sum with short, so an upcast (Convert) it introduced on the parameter to int. Do you see anything missing or a problem here?

Reply:
Ah, I see, no problem.